### PR TITLE
Syncthing Chart

### PR DESCRIPTION
**Pull request by marshallford**
#### What this PR does / why we need it:

Adds Syncthing chart

#### Checklist
- [X] [DCO](https://www.helm.sh/blog/helm-dco/index.html) signed
- [X] Variables are documented in the README.md

### CODE REVIEW

File: incubator/syncthing/.helmignore [hunk 1]
```diff
@@ -0,0 +1,22 @@
+# Patterns to ignore when building packages.
```
Comment (davidkarlsen) [inline incubator/syncthing/.helmignore]: go straight for the stable repo

Comment (davidkarlsen) [reply incubator/syncthing/.helmignore]: Add use of serviceAccount, see: https://github.com/helm/helm/blob/master/docs/chart_best_practices/rbac.md for howto

File: incubator/syncthing/Chart.yaml [hunk 2]
```diff
@@ -0,0 +1,16 @@
+apiVersion: v1
+appVersion: "137"
+description: Open Source Continuous File Synchronization
+name: syncthing
+home: https://syncthing.net/
+version: 0.1.0
```
Comment (davidkarlsen) [inline incubator/syncthing/Chart.yaml]: start with version 1.0.0

**Review by davidkarlsen — CHANGES_REQUESTED** [review]

File: stable/syncthing/templates/ingress.yaml [hunk 3]
```diff
@@ -0,0 +1,38 @@
+{{- if .Values.ingress.enabled -}}
+{{- $fullName := include "syncthing.fullname" . -}}
+{{- $ingressPath := .Values.ingress.path -}}
+apiVersion: extensions/v1beta1
+kind: Ingress
+metadata:
+  name: {{ $fullName }}
+  labels:
+    app.kubernetes.io/name: {{ include "syncthing.name" . }}
```
Comment (davidkarlsen) [inline stable/syncthing/templates/ingress.yaml]: these label names differ from the other ones used.
There was a switch in the best-practice naming - but please use a consistent set throughout the chart.

**Review by davidkarlsen — CHANGES_REQUESTED** [review]

File: stable/syncthing/templates/ingress.yaml (hunk 3 above)
Comment (marshallford) [reply stable/syncthing/templates/ingress.yaml]: Fixed.

File: stable/syncthing/templates/discovery-svc.yaml [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+  ports:
+    - port: {{ .Values.discoveryService.port }}
+      targetPort: discovery
+      protocol: UDP
+      name: discovery
+{{ if (and (eq .Values.discoveryService.type "NodePort") (not (empty .Values.discoveryService.nodePort))) }}
+      nodePort: {{.Values.discoveryService.nodePort}}
+{{ end }}
+  selector:
+    app: {{ template "syncthing.name" . }}
```
Comment (benosman) [inline stable/syncthing/templates/discovery-svc.yaml]:
```suggestion
    app.kubernetes.io/name: {{ template "syncthing.name" . }}
```

File: stable/syncthing/templates/listener-svc.yaml [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+  ports:
+    - port: {{ .Values.listenerService.port }}
+      targetPort: listener
+      protocol: TCP
+      name: listener
+{{ if (and (eq .Values.listenerService.type "NodePort") (not (empty .Values.listenerService.nodePort))) }}
+      nodePort: {{.Values.listenerService.nodePort}}
+{{ end }}
+  selector:
+    app: {{ template "syncthing.name" . }}
```
Comment (benosman) [inline stable/syncthing/templates/listener-svc.yaml:51]:
```suggestion
    app.kubernetes.io/name: {{ template "syncthing.name" . }}
```

File: stable/syncthing/templates/webui-svc.yaml [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+  ports:
+    - port: {{ .Values.webuiService.port }}
+      targetPort: webui
+      protocol: TCP
+      name: webui
+{{ if (and (eq .Values.webuiService.type "NodePort") (not (empty .Values.webuiService.nodePort))) }}
+      nodePort: {{.Values.webuiService.nodePort}}
+{{ end }}
+  selector:
+    app: {{ template "syncthing.name" . }}
```
Comment (benosman) [inline stable/syncthing/templates/webui-svc.yaml:51]:
```suggestion
    app.kubernetes.io/name: {{ template "syncthing.name" . }}
```

File: stable/syncthing/templates/discovery-svc.yaml [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+    release: {{ .Release.Name }}
```
Comment (benosman) [inline stable/syncthing/templates/discovery-svc.yaml]:
```suggestion
    app.kubernetes.io/instance: {{ .Release.Name }}
```

File: stable/syncthing/templates/listener-svc.yaml [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+    release: {{ .Release.Name }}
```
Comment (benosman) [inline stable/syncthing/templates/listener-svc.yaml:52]:
```suggestion
    app.kubernetes.io/instance: {{ .Release.Name }}
```

File: stable/syncthing/templates/webui-svc.yaml [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+    release: {{ .Release.Name }}
```
Comment (benosman) [inline stable/syncthing/templates/webui-svc.yaml:52]:
```suggestion
    app.kubernetes.io/instance: {{ .Release.Name }}
```

File: stable/syncthing/templates/NOTES.txt [hunk 10]
```diff
@@ -0,0 +1,20 @@
...
+  export NODE_PORT=$(kubectl get --namespace {{ .Release.Namespace }} -o jsonpath="{.spec.ports[0].nodePort}" services {{ template "syncthing.fullname" . }})
+  export NODE_IP=$(kubectl get nodes --namespace {{ .Release.Namespace }} -o jsonpath="{.items[0].status.addresses[0].address}")
+  echo http://$NODE_IP:$NODE_PORT
+{{- else if contains "LoadBalancer" .Values.webuiService.type }}
+     NOTE: It may take a few minutes for the LoadBalancer IP to be available.
+           You can watch the status of by running 'kubectl get svc -w {{ template "syncthing.fullname" . }}'
+  export SERVICE_IP=$(kubectl get svc --namespace {{ .Release.Namespace }} {{ template "syncthing.fullname" . }} -o jsonpath='{.status.loadBalancer.ingress[0].ip}')
+  echo http://$SERVICE_IP:{{ .Values.webuiService.port }}
+{{- else if contains "ClusterIP" .Values.webuiService.type }}
+  export POD_NAME=$(kubectl get pods --namespace {{ .Release.Namespace }} -l "app={{ template "syncthing.name" . }},release={{ .Release.Name }}" -o jsonpath="{.items[0].metadata.name}")
```
Comment (benosman) [inline stable/syncthing/templates/NOTES.txt:16]:
```suggestion
  export POD_NAME=$(kubectl get pods --namespace {{ .Release.Namespace }} -l "app.kubernetes.io/name={{ template "syncthing.name" . }},app.kubernetes.io/instance={{ .Release.Name }}" -o jsonpath="{.items[0].metadata.name}")
```

**Review by benosman — CHANGES_REQUESTED** [review]
The labels follow the newer helm/kubernetes standards such as using app.kubernetes.io/name instead of app, but the selectors in the services still reference the old style.

Additionally, NOTES.txt also references the old labels in the example commands.
 
Without these changes the services do not have the endpoints set correctly.

File: stable/syncthing/templates/deployment.yaml [hunk 11]
```diff
@@ -0,0 +1,71 @@
...
+            - name: webui
+              containerPort: 8384
+              protocol: TCP
+            - name: listener
+              containerPort: 22000
+              protocol: TCP
+            - name: discovery
+              containerPort: 21027
+              protocol: UDP
+          livenessProbe:
```
Comment (benosman) [inline stable/syncthing/templates/deployment.yaml:37]: I'm experiencing issues with the pod constantly restarting, without any obvious error message. I get the feeling that the probes are causing this, so ideally these settings should be customizable as well as less aggressive by default.

It also looks like the indentation on  `initialDelaySeconds` is incorrect, grouping it under httpGet.

**Update:** I can confirm that changing the period of the probes and increasing the failure count improved things, but didn't fully resolve the restarting. Removing the probes altogether **did** stop the restarting. Now the uptime is well over 16 hours and it's still responding fine. 

It seems like the UI often becomes unresponsive for short periods of time, even when the daemon itself is fine. I wonder if there is a better way to healthcheck.

**Comment by cbenhagen** [comment]
@marshallford What is the reason for not using the [official images](https://hub.docker.com/r/syncthing/syncthing/)?

**Comment by marshallford** [comment]
@cbenhagen The `linuxserver/syncthing` version seemed to be more popular but I will certainly look into the official image as well. Apologies for the slow work, KubeCon + Holidays has slowed me down quite a bit.

**Comment by nrvale0** [comment]
I'd like to be able to deploy syncthing on k8s and feel the easiest path might be to continue this work. Before I start, would I need to create my own fork of the original author's repo and submit my own PR? Or?

**Pull request closed without merging**